### PR TITLE
[WIP][Feature] Support TransFusion

### PR DESCRIPTION
**Pull request by wep21**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

import [TransFusion](https://github.com/XuyangBai/TransFusion) as project

## Modification

Please briefly describe what modification is made in this PR.
add transfusion implementation

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by JingweiZhang12** [comment]
@wep21 Thanks for your awesome contribution, and we're excited to invite you to join **core developer community** of mmdet3d. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID". We will invite you to MMDet3D developer group.

File: projects/TransFusion/configs/transfusion_nusc_pillar_L.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
+_base_ = ['mmdet3d::_base_/default_runtime.py']
```
Comment (JingweiZhang12) [inline projects/TransFusion/configs/transfusion_nusc_pillar_L.py:1]:
```suggestion
_base_ = ['../../../configs/_base_/default_runtime.py']
```
Better to use a relative path to avoid potential problems with openmim.

File: projects/TransFusion/transfusion/hungarian_assigner.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    def __init__(self, weight):
+        self.weight = weight
+
+    def __call__(self, iou):
+        iou_cost = -iou
+        return iou_cost * self.weight
+
+
+@TASK_UTILS.register_module()
+class HeuristicAssigner3D(BaseAssigner):
```
Comment (JingweiZhang12) [inline projects/TransFusion/transfusion/hungarian_assigner.py:55]: Could reuse the implementation in BEVFusion and del this file?
```python
from ..BEVFusion.bevfusion.utils import HungarianAssigner3D
from ..BEVFusion.bevfusion.transformer import TransformerDecoderLayer  # noqa F401
from ..BEVFusion.bevfusion.utils import TransFusionBBoxCoder
```

File: projects/TransFusion/transfusion/transfusion_bbox_corder.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
+import torch
+from mmdet.models.task_modules import BaseBBoxCoder
+
+from mmdet3d.registry import TASK_UTILS
+
+
+@TASK_UTILS.register_module()
+class TransFusionBBoxCoder(BaseBBoxCoder):
```
Comment (JingweiZhang12) [inline projects/TransFusion/transfusion/transfusion_bbox_corder.py:8]: The same as the above

**Comment by JingweiZhang12** [comment]
Could you update your progress of implementation of TransFusion  according to the bellow items:
- [ ] Support inference
- [ ] Align the precision of inference
- [ ] Support training
- [ ] Align the precision of training
- [ ] Provide docs

File: projects/TransFusion/transfusion/hungarian_assigner.py (hunk 2 above)
Comment (wep21) [reply projects/TransFusion/transfusion/hungarian_assigner.py:55]: Sure. Is it better to merge [these changes](https://github.com/open-mmlab/mmdetection3d/pull/2558)?

Comment (wep21) [reply projects/TransFusion/transfusion/hungarian_assigner.py:51]: @JingweiZhang12 I pushed the code based on the original repository. I'm currently working on using BEVFusion implementation [here](https://github.com/wep21/mmdetection3d/tree/transfusion-with-bevfusion), but loss_bbox doesn't decrease compared to the original one.
I guess I made a mistake when integrating your code, so I appreciate it if you could give me some advice to resolve this issue.

Comment (JingweiZhang12) [reply projects/TransFusion/transfusion/hungarian_assigner.py:51]: @wep21 Hi, the original BEVFusion has a potential bug: https://github.com/mit-han-lab/bevfusion/issues/419. This will cause the loss_bbox can not decrease. We have fixed it in the latest branch.

Comment (wep21) [reply projects/TransFusion/transfusion/hungarian_assigner.py]: @JingweiZhang12 I found out the changes from original code cause problem when training transfusion. I reverted the changes [here](https://github.com/open-mmlab/mmdetection3d/pull/2547/commits/2ad5e4ac9cbbfc2a93a99ca85983d57d7e05b450) and then loss_bbox and matched ious become sane.
Do you have any thoughts about this?

Comment (JingweiZhang12) [reply projects/TransFusion/transfusion/hungarian_assigner.py]: Hi， this is because the voxelization output order of the transfusion is different from ours. TransFusion: zyx,   Ours: zxy

Comment (JingweiZhang12) [reply projects/TransFusion/transfusion/hungarian_assigner.py]: It may not have an influence on the accuracy of the model. The style of the original Transfusion is the same as ours.

Comment (wep21) [reply projects/TransFusion/transfusion/hungarian_assigner.py]: @JingweiZhang12 I would like to use the same coordinate as original, so is it better to create another head implementation in transfusion?

File: projects/BEVFusion/bevfusion/transfusion_head.py [hunk 4]
```diff
@@ -725,10 +724,11 @@ def get_targets_single(self, gt_instances_3d, preds_dict, batch_idx):
                 center_int = center.to(torch.int32)
 
                 # original
-                # draw_heatmap_gaussian(heatmap[gt_labels_3d[idx]], center_int, radius) # noqa: E501
+                draw_heatmap_gaussian(heatmap[gt_labels_3d[idx]], center_int,
+                                      radius)  # noqa: E501
                 # NOTE: fix
-                draw_heatmap_gaussian(heatmap[gt_labels_3d[idx]],
-                                      center_int[[1, 0]], radius)
+                # draw_heatmap_gaussian(heatmap[gt_labels_3d[idx]],
```
Comment (JingweiZhang12) [inline projects/BEVFusion/bevfusion/transfusion_head.py:730]: Please remove lines 729-731 since it's unnecessary.

File: projects/TransFusion/configs/transfusion_nusc_pillar_L.py [hunk 5]
```diff
@@ -0,0 +1,397 @@
+_base_ = ['../../../configs/_base_/default_runtime.py']
+custom_imports = dict(
+    imports=['projects.TransFusion.transfusion'], allow_failed_imports=False)
+point_cloud_range = [-51.2, -51.2, -5.0, 51.2, 51.2, 3.0]
```
Comment (JingweiZhang12) [inline projects/TransFusion/configs/transfusion_nusc_pillar_L.py:4]: Please rename this file as `transfusion_lidar_pillar02_second_secfpn_8xb2-cyclic-20e_nus-3d.py`

File: projects/TransFusion/configs/transfusion_nusc_pillar_L.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+input_modality = dict(
+    use_lidar=True,
+    use_camera=False,
+    use_radar=False,
+    use_map=False,
+    use_external=False)
+backend_args = None
+
+model = dict(
+    type='TransFusionDetector',
```
Comment (JingweiZhang12) [inline projects/TransFusion/configs/transfusion_nusc_pillar_L.py:41]:
```suggestion
    type='TransFusion',
```
Remind to modify the corresponding name.

File: projects/TransFusion/transfusion/hungarian_assigner.py (hunk 2 above)
Comment (JingweiZhang12) [reply projects/TransFusion/transfusion/hungarian_assigner.py]: @wep21 Yep, it's better to use the original head implements of TransFusioin to align the training precision easily since the BEVFusion head may modify something and have some potential incompatibilities.

**Comment by JingweiZhang12** [comment]
@wep21 Hi, have you aligned the inference precision of TransFusion(only LiDAR modality)?

File: projects/TransFusion/configs/transfusion_nusc_pillar_L.py (hunk 5 above)
Comment (wep21) [reply projects/TransFusion/configs/transfusion_nusc_pillar_L.py]: renamed config at https://github.com/open-mmlab/mmdetection3d/pull/2547/commits/615cb1d6fa1bd3f1dc3876a5120cf2b8ffdf534d.

File: projects/TransFusion/configs/transfusion_nusc_pillar_L.py (hunk 6 above)
Comment (wep21) [reply projects/TransFusion/configs/transfusion_nusc_pillar_L.py]: rename detector class name at https://github.com/open-mmlab/mmdetection3d/pull/2547/commits/5608d19879107b656c0276185801e1a46747305d

File: projects/TransFusion/transfusion/transfusion.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
+from mmdet3d.models.detectors.mvx_two_stage import MVXTwoStageDetector
+from mmdet3d.registry import MODELS
+
+
+@MODELS.register_module()
+class TransFusionDetector(MVXTwoStageDetector):
+    """Base class of Multi-modality VoxelNet."""
```
[the remark this answers is not in the document]
Comment (wep21) [reply projects/TransFusion/transfusion/transfusion.py]: updated at https://github.com/open-mmlab/mmdetection3d/pull/2547/commits/fd47bf7b7b383fd97c32d6679d387d4ff7924acf

File: projects/BEVFusion/bevfusion/transfusion_head.py (hunk 4 above)
Comment (wep21) [reply projects/BEVFusion/bevfusion/transfusion_head.py]: remove these lines in new transfusion head at https://github.com/open-mmlab/mmdetection3d/pull/2547/commits/bb0a95e7be5c4eaaf318d6af3f07f40c370c1cca.
Also, I renamed head in bevfusion due to name conflict at https://github.com/open-mmlab/mmdetection3d/pull/2547/commits/dc948cc88b7730974eb1f0550cd37e2f36bf244b. Please let me know if this change is not applicable.

**Comment by wep21** [comment]
@JingweiZhang12 

> Hi, have you aligned the inference precision of TransFusion(only LiDAR modality)?

I'm sorry, but not yet. I've tested with only nuscenes-mini. Currently, I don't have enough gpu environment, it may take a time to train and test with nuscenes full dataset.

**Comment by JingweiZhang12** [comment]
> @JingweiZhang12
> 
> > Hi, have you aligned the inference precision of TransFusion(only LiDAR modality)?
> 
> I'm sorry, but not yet. I've tested with only nuscenes-mini. Currently, I don't have enough gpu environment, it may take a time to train and test with nuscenes full dataset.

That's OK. Suggest you can make sure the results of one forward are the same as that of the official implementation firstly.

**Comment by sunjiahao1999** [comment]
@wep21 Hi, any progress on TransFusion so far?

**Comment by wep21** [comment]
@sunjiahao1999 I'm sorry for late reply. This is the evaluation result with nuscenes dataset.
```
Evaluating bboxes of pred_instances_3d
mAP: 0.5562                                                                                                                                                                                                                                                                  
mATE: 0.3236
mASE: 0.2586
mAOE: 0.3608
mAVE: 0.3053
mAAE: 0.1887
NDS: 0.6344
```

**Comment by wep21** [comment]
[20230819_092709.log](https://github.com/open-mmlab/mmdetection3d/files/12421755/20230819_092709.log)

**Comment by wep21** [comment]
image
This figure comes from the original paper of transfusion. The second row shows the score in the same condition as my experiment.

**Comment by wep21** [comment]
inference demo
```
python demo/pcd_demo.py demo/data/nuscenes/n015-2018-07-24-11-22-45+0800__LIDAR_TOP__1532402927647951.pcd.bin projects/TransFusion/configs/transfusion_lidar_pillar02_second_secfpn_8xb2-cyclic-20e_nus-3d.py ~/Downloads/mmdet3d_epoch20.pth --show --score-thr 0.05
```
image

**Comment by JingweiZhang12** [comment]
@wep21 Please fix the lint.

**Comment by wep21** [comment]
@JingweiZhang12 I fixed lint at https://github.com/open-mmlab/mmdetection3d/pull/2547/commits/2cb5f7876fcb58a5ce0182581a85174056265b3c.

**Comment by wep21** [comment]
@JingweiZhang12 @sunjiahao1999 friendly ping

**Comment by friendship1** [comment]
I was looking for this transfusion code. Thank you.

Just to double-check, are the 54.4 and 64.7 values in the table in the paper you posted (https://github.com/open-mmlab/mmdetection3d/pull/2547#issuecomment-1690360882), 2nd row, left column, the ones you're comparing to (https://github.com/open-mmlab/mmdetection3d/pull/2547#issuecomment-1690340922)? In other words, is it correct that your experiment yielded higher numbers than the paper (just for pillar based, lidar-only)?

Also, I know that the original Transfusion paper used a "fading strategy", and I was wondering if this was applied in your code or in your experiments.

Thanks the great work!

**Comment by friendship1** [comment]
I just checked that the fading strategy is implemented as a custom hook. 

But what you implemented doesn't seem to use a 2-stage learning method: pretrain only the LiDAR model first and then further train this model on additional modality with camera input added.
is that correct?
